src/NServiceBus.Core/Serializers/SystemJson/JsonMessageSerializer.cs

```diff
@@ -0,0 +1,83 @@
+﻿namespace NServiceBus.Serializers.SystemJson
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Text.Json;
+    using NServiceBus.Serialization;
+
+    class JsonMessageSerializer : IMessageSerializer
+    {
+        public JsonMessageSerializer(JsonSerializerOptions serializerOptions, JsonWriterOptions writerOptions, JsonReaderOptions readerOptions, string contentType)
+        {
+            this.serializerOptions = serializerOptions;
+            this.writerOptions = writerOptions;
+            this.readerOptions = readerOptions;
+
+            if (contentType == null)
+            {
+                ContentType = ContentTypes.Json;
+            }
+            else
+            {
+                ContentType = contentType;
+            }
+        }
+
+        public void Serialize(object message, Stream stream)
+        {
+            using var writer = new Utf8JsonWriter(stream, writerOptions);
+            JsonSerializer.Serialize(writer, message, serializerOptions);
+        }
+
+        public object[] Deserialize(ReadOnlyMemory<byte> body, IList<Type> messageTypes = null)
+        {
+            if (messageTypes == null || !messageTypes.Any())
+            {
+                throw new("The System.Text.Json message serializer requires message types to be defined.");
+            }
+
+            if (messageTypes.Count == 1)
+            {
+                return new[] { Deserialize(body, messageTypes[0]) };
+            }
+
+            var rootTypes = FindRootTypes(messageTypes);
+            return rootTypes.Select(rootType => Deserialize(body, rootType))
+                .ToArray();
+        }
+
+        object Deserialize(ReadOnlyMemory<byte> body, Type type)
+        {
+            var reader = new Utf8JsonReader(body.Span, readerOptions);
+            return JsonSerializer.Deserialize(ref reader, type, serializerOptions)!;
+        }
+
+        static IEnumerable<Type> FindRootTypes(IEnumerable<Type> messageTypesToDeserialize)
```

Comment:
Philosophical remark. Sad that all the serializers have to have this logic while the deserialize connector already some complicated enclosed message type extraction, caching and handling there. 

https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Pipeline/Incoming/DeserializeMessageConnector.cs#L72

Reply:
Sounds like something that could/should be tackled when the UseSerialization API is matched up to the transport API?

Reply:
Agreed